api/src/main/java/org/xnio/XnioWorker.java

```diff
@@ -128,7 +128,12 @@ public void run() {
         };
         final ExecutorService executorService = builder.getExternalExecutorService();
         if (executorService != null) {
-            taskPool = new ExternalTaskPool(executorService);
+            if (executorService instanceof EnhancedQueueExecutor) {
+                taskPool = new ExternalTaskPool(
+                        new EnhancedQueueExecutorTaskPool((EnhancedQueueExecutor) executorService));
+            } else {
+                taskPool = new ExternalTaskPool(new ExecutorServiceTaskPool(executorService));
+            }
```

Comment:
If we've gone this far, it almost feels like we should be adapting ThreadPoolExecutors as well.  It's just that, as a matter of expediency, the TPE TaskPool implements that interface directly and subclasses TPE, so it can't really be reused...

Reply:
Good call. I've split the ThreadPoolExecutorTaskPool into `DefaultThreadPoolExecutor` with the custom shutdown task and custom resize logic, and a `ThreadPoolExecutorTaskPool` which delegates to a delegate `ThreadPoolExecutor`.